src/file/format/ini.rs

```diff
@@ -3,35 +3,45 @@ use std::error::Error;
 use ini::Ini;
 
 use crate::map::Map;
-use crate::value::{Value, ValueKind};
+use crate::value::{Table, Value, ValueKind};
 
 pub fn parse(
     uri: Option<&String>,
     text: &str,
 ) -> Result<Map<String, Value>, Box<dyn Error + Send + Sync>> {
-    let mut map: Map<String, Value> = Map::new();
-    let i = Ini::load_from_str(text)?;
-    for (sec, prop) in i.iter() {
-        match sec {
-            Some(sec) => {
-                let mut sec_map: Map<String, Value> = Map::new();
-                for (k, v) in prop.iter() {
-                    sec_map.insert(
-                        k.to_owned(),
-                        Value::new(uri, ValueKind::String(v.to_owned())),
-                    );
-                }
-                map.insert(sec.to_owned(), Value::new(uri, ValueKind::Table(sec_map)));
-            }
-            None => {
-                for (k, v) in prop.iter() {
-                    map.insert(
-                        k.to_owned(),
-                        Value::new(uri, ValueKind::String(v.to_owned())),
-                    );
-                }
-            }
-        }
+    let value = from_ini(uri, Ini::load_from_str(text)?);
+
+    match value.kind {
+        ValueKind::Table(map) => Ok(map),
+
+        _ => Ok(Map::new()),
     }
-    Ok(map)
+}
+
+fn from_ini(
+    uri: Option<&String>,
+    data: Ini,
+) -> Value {
+    let mut map = Map::<String, Value>::new();
+
+    let mut sections: Map<Option<&str>, Table> = data.into_iter().map(|(section, props)| {(
+        section,
+        props.iter().map(|(k, v)| {
+            let key = k.to_owned();
+            let value = Value::new(uri, ValueKind::String(v.to_owned()));
+            (key, value)
+        }).collect()
+    )}).collect();
```

Comment:
Here we have an iterator inside an iterator. I bet this could be rewritten to something like:

```suggestion
    let mut sections: Map<Option<&str>, Table> = data.into_iter().flat_map(|(section, props)| {(
        section,
        props.iter().map(|(k, v)| {
            let key = k.to_owned();
            let value = Value::new(uri, ValueKind::String(v.to_owned()));
            (key, value)
        })
    )}).collect();
```

Also I don't like the outer tuple (didn't notice before), so that could be expanded as well.

Sorry for continuously bugging you about this piece :laughing:

Reply:
> Here we have an iterator inside an iterator. I bet this could be rewritten to something like:

I don't think [FlatMap](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flat_map) is going to work. We are collecting like this:

```
sectionless_property=value

[section_a]
property_in_section=value
```

Each section has it's own collection of properties, while the config is a collection of sections (_with the sectionless one hoisted up_). The nesting is necessary? At an earlier iteration, I had extracted the inner collection for properties to a separate function call, I could switch to that if you find it more readable.

> Also I don't like the outer tuple (didn't notice before), so that could be expanded as well.

No worries, I'll take care of that 👍 

> Sorry for continuously bugging you about this piece 😆

It's alright, I'm the same when I review contributions too 😛

Reply:
> I don't think [FlatMap](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flat_map) is going to work. We are collecting like this:

Ah, now I see it that there's a tuple as well! That's an oversight at my side, sorry. Sure, then `flat_map` does not work here.